feat(hy2): implement salamander obfuscation

ProtoFN opened the pull request:
Add Salamander obfuscation implementation for Hysteria2.

It's my first Go experience. As I know there is no need for array pooling, check it if is not true. Also it increases UDP packet size, so can make problems with MTU. Also I think that XORing can be SIMDed and hash of first part (that contains key) can be precomputed.

It is working on my setup. Tests are broken so can't run them.

Closes https://github.com/daeuniverse/dae/issues/904
Closes https://github.com/daeuniverse/dae/issues/798

FireNameFN committed:
feat(hy2): implement salamander obfuscation

AkinoKaede commented on protocol/hysteria2/client/client.go:
```diff
@@ -63,6 +63,10 @@ func (c *clientImpl) connect(ctx context.Context) (*HandshakeInfo, error) {
 	if err != nil {
 		return nil, err
 	}
+	// Wrap in Salamander implementation
```
From a code robustness perspective, it's best to return an error for illegal values.

AkinoKaede committed:
feat: obfuscation mode validation

Prevents unknown obfuscation values by validating the mode
and returning an error for unsupported options, improving
configuration safety and error transparency.

ProtoFN commented on protocol/hysteria2/client/client.go:
```diff
@@ -63,10 +63,15 @@ func (c *clientImpl) connect(ctx context.Context) (*HandshakeInfo, error) {
-	// Wrap in Salamander implementation
-	if c.config.ObfuscationConfig.Obfuscation == "salamander" {
+
+	switch c.config.ObfuscationConfig.Obfuscation {
+	case "salamander":
 		pktConn = NewSalamanderPacketConn(pktConn, c.config.ObfuscationConfig.ObfuscationKey)
+	case "":
+	default:
+		return nil, errors.New("unknown obfuscation: " + c.config.ObfuscationConfig.Obfuscation)
 	}
```
In [Hysteria2 source code](https://github.com/apernet/hysteria/blob/44a5643535bf63760659b2a8c76fde6330792ab8/app/cmd/server.go#L276) `plain` also works as without obfuscation.

AkinoKaede replied:
Thanks for your reminder.

FireNameFN commented:
I don't forget about this PR, just was busy.

AkinoKaede committed:
fix: `plain` as obfuscation method alias

ProtoFN commented on dialer/hysteria2/hysteria2.go:
Maybe remove this line?

AkinoKaede replied:
Yes, u r right.

FireNameFN committed:
style: remove obsolete todo comment
